src/ray/raylet/worker_pool.cc

```diff
@@ -85,9 +106,29 @@ WorkerPool::WorkerPool(int num_workers, int maximum_startup_concurrency,
     state.worker_command = entry.second;
     RAY_CHECK(!state.worker_command.empty()) << "Worker command must not be empty.";
   }
+  signals_.async_wait(boost::bind(&WorkerPool::HandleSIGCHLD, this, _1, _2));
   Start(num_workers);
 }
 
+void WorkerPool::HandleSIGCHLD(const boost::system::error_code &error,
+                               int signal_number) {
+  if (!error) {
+    pid_t child_pid;
+    while ((child_pid = waitpid(-1, NULL, WNOHANG)) > 0) {
+      for (auto &entry : states_by_lang_) {
+        auto &state = entry.second;
+        auto it = state.pid_to_connections.find(child_pid);
+        if (it != state.pid_to_connections.end()) {
+          for (const auto &client : it->second) {
+            worker_death_callback_(client);
+          }
+        }
+      }
+    }
+  }
+  signals_.async_wait(boost::bind(&WorkerPool::HandleSIGCHLD, this, _1, _2));
```

Comment:
Is it possible that some signals are lost after the last `waitpid`, but before this `async_wait` is called?

Reply:
They will be queued when there isn't a waiting handler:
https://www.boost.org/doc/libs/1_47_0/doc/html/boost_asio/reference/signal_set.html